### PR TITLE
Image segmentation YOLOv8

### PR DESCRIPTION
**Pull request by duskovesic**

### CODE REVIEW

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    datasetPath: Path,
+    trainDatasetPath: Path,
+    validDatasetPath: Path,
+    classes: ImageDatasetClasses,
+    location: Path
+) -> None:
+
+    classesIds: dict[int, str] = {}
+    for index, clazz in enumerate(classes):
+        classesIds[index] = clazz.label
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:35]: It might make more sense to rename this to `classesCategorical`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    return True
+
+
+def createYamlFile(
+    datasetPath: Path,
+    trainDatasetPath: Path,
+    validDatasetPath: Path,
+    classes: ImageDatasetClasses,
+    location: Path
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:30]: Let's rename this to `path`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        classesIds[index] = clazz.label
+
+    data = {
+        "path": str(datasetPath.absolute()),
+        "train": str(trainDatasetPath.relative_to(datasetPath)),
+        "val": str(validDatasetPath.relative_to(datasetPath)),
+        "names": classesIds
+    }
+
+    with open(location, "w") as yamlFIle:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:44]: Use `pathlib.Path.open` instead of `open`

Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:44]: `yamlFIle` -> `yamlFile`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        yaml.dump(data, yamlFIle, default_flow_style = False, sort_keys = False)
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:45]: Use `yaml.safe_dump` instead of `dump`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+def loadAnnotation(sample: ImageSample) -> Optional[CoretexImageAnnotation]:
+    if not sample.annotationPath.exists():
+        return None
+
+    with sample.annotationPath.open("r") as file:
+        return CoretexImageAnnotation.decode(json.load(file))
+
+
+def addImagePathForYaml(sample: ImageSample, imagesPath: Path) -> None:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:56]: There is no need for this to be a separate function

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    sample.unzip()
+    sample.imagePath.link_to(imagesPath / f"{sample.id}{sample.imagePath.suffix}")
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:58]: `
destinationPath = imagesPath.joinpath(str(sample.id)).with_suffix(sample.imagePath.suffix)
sample.imagePath.link_to(destinationPath)
`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            for segmentation in instance.segmentations:
+                w = annotation.width
+                h = annotation.height
+                normalizedSegmentation = [num / w if idx % 2 == 0 else num / h for idx, num in enumerate(segmentation)]
+
+                normalizedAnnotation = labelId, normalizedSegmentation
+
+                normalizedAnnotations.append(normalizedAnnotation)
+
+        with open(location / f"{sample.id}.txt", "w") as txtFile:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:83]: Use `pathlib.Path.open` instead of open

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        for instance in annotation.instances:
+            labelId = classes.labelIdForClassId(instance.classId)
+
+            if labelId is None:
+                continue
+
+            for segmentation in instance.segmentations:
+                w = annotation.width
+                h = annotation.height
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:76]: Use full names: `width` and `height`

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                normalizedSegmentation = [num / w if idx % 2 == 0 else num / h for idx, num in enumerate(segmentation)]
+
+                normalizedAnnotation = labelId, normalizedSegmentation
+
+                normalizedAnnotations.append(normalizedAnnotation)
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:81]: There is no need to create `normalizedAnnotation` variable as it is only appended to the list. Append the values directly.

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+        with open(location / f"{sample.id}.txt", "w") as txtFile:
+            for segmentation in normalizedAnnotations:
+                txtFile.write(str(segmentation[0]))
+                coordinates = segmentation[1]
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:86]: For what is this variable used?

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                h = annotation.height
+                normalizedSegmentation = [num / w if idx % 2 == 0 else num / h for idx, num in enumerate(segmentation)]
+
+                normalizedAnnotation = labelId, normalizedSegmentation
+
+                normalizedAnnotations.append(normalizedAnnotation)
+
+        with open(location / f"{sample.id}.txt", "w") as txtFile:
+            for segmentation in normalizedAnnotations:
+                txtFile.write(str(segmentation[0]))
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:85]: Why is this written to file separately?

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+def addAnnotation(sample: ImageSample, classes: ImageDatasetClasses, location: Path) -> None:
+    sample.unzip()
+
+    annotation = loadAnnotation(sample)
+    if annotation is not None:
+        normalizedAnnotations: list[tuple[int, list[float]]] = []
+
+        for instance in annotation.instances:
+            labelId = classes.labelIdForClassId(instance.classId)
+
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:70]: Remove this newline

File: tasks/image-segmentation-yolov8/src/dataset.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    with sample.annotationPath.open("r") as file:
+        return CoretexImageAnnotation.decode(json.load(file))
+
+
+def addImagePathForYaml(sample: ImageSample, imagesPath: Path) -> None:
+    sample.unzip()
+    sample.imagePath.link_to(imagesPath / f"{sample.id}{sample.imagePath.suffix}")
+
+
+def addAnnotation(sample: ImageSample, classes: ImageDatasetClasses, location: Path) -> None:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:61]: `location` -> `path`

Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/dataset.py:61]: `addAnnotation` -> `saveYoloAnnotation`

File: tasks/image-segmentation-yolov8/src/train.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    if not isValidationSplitValid(taskRun.parameters["validationSplit"], taskRun.dataset.count):
+        raise ValueError(f">> [ObjectDetection] validationSplit parameter is invalid")
+
+    datasetPath = folder_manager.createTempFolder("dataset")
+    trainDatasetPath, validDatasetPath = prepareDataset(taskRun.dataset, datasetPath, taskRun.parameters["validationSplit"])
+
+    yamlFilePath = datasetPath / "config.yaml"
+    createYamlFile(datasetPath, trainDatasetPath, validDatasetPath, taskRun.dataset.classes, yamlFilePath)
+
+    model = YOLO("yolov8n-seg.pt")
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/train.py:36]: Weights are hardcoded, let's extract this into an enum parameter same as YOLOv8 detection

File: tasks/image-segmentation-yolov8/src/train.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    model.train(
+        data = yamlFilePath,
+        epochs = taskRun.parameters["epochs"],
+        batch = taskRun.parameters["batchSize"],
+        imgsz = taskRun.parameters["imageSize"],
+        patience = getPatience(taskRun)
+    )
+
+    logging.info(">> [Image Segmentation] Validating the model")
+    model.val()
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/train.py:51]: Are results of this unused?

File: tasks/image-segmentation-yolov8/src/train.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    accuracy = validate(taskRun)
+
+    modelName = taskRun.generateEntityName()
+    onnxModelPath = model.export(format = "onnx")
+    ctxModel = Model.createModel(modelName, taskRun.projectId, accuracy / 100, {})
+    ctxModel.upload(Path(onnxModelPath).parent)
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/train.py:58]: Why is only onnx model being uploaded?

Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/train.py:58]: `model_descriptor.json` should be uploaded as well with information about the model input (image size) and model output (classes)

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+    return csvClassesData
+
+
+def generateFieldNamesDataset(labels: list[str]) -> list[str]:
+    fieldNamesDataset: list[str] = []
+
+    for label in labels:
+        fieldNamesDataset.append(f"{label} acc")
+        fieldNamesDataset.append(f'{label} acc STD')
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:43]: Labels should not have separate columns, but rows.

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+    fieldNamesDataset = generateFieldNamesDataset(taskRun.dataset.classes.labels)
+
+    csvDatasetPath = folder_manager.createTempFolder("csv_dataset") / "dataset_results.csv"
+
+    csvClassesData = fetchCsvClassesData(taskRun.dataset.classes.labels, csvSamplesPath)
+    csvDatasetAcc = fetchCsvDatasetAcc(csvSamplesPath)
+
+    csvDatasetResults: dict[str, str] = {}
+    for label in taskRun.dataset.classes.labels:
+        csvDatasetResults[f"{label} acc"] = f"{np.mean(csvClassesData[label]):.2f}"
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:60]: Use `Accuracy` and `Accuracy stdev` as column names everywhere to keep it consistent.

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+    fieldNamesDataset.extend(["accuracy", "accuracy STD"])
+
+    return fieldNamesDataset
+
+
+def createCsvDatasetResults(taskRun: TaskRun[ImageDataset], csvSamplesPath: Path) -> float:
+    fieldNamesDataset = generateFieldNamesDataset(taskRun.dataset.classes.labels)
+
+    csvDatasetPath = folder_manager.createTempFolder("csv_dataset") / "dataset_results.csv"
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:53]: `datasetResultsPath`

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+
+    csvClassesData = fetchCsvClassesData(taskRun.dataset.classes.labels, csvSamplesPath)
+    csvDatasetAcc = fetchCsvDatasetAcc(csvSamplesPath)
+
+    csvDatasetResults: dict[str, str] = {}
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:58]: `datasetResult`

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+
+            iouClass = iouScoreClass(sample.load().annotation.extractSegmentationMask(ImageDatasetClasses([clazz])), decodedClassMask)
+            iouClasses.append(iouClass)
+            csvSampleValues.append(f"{iouClass * 100:.2f}")
+
+        if len(iouClasses) > 0:
+            sampleAcc = sum(iouClasses) / len(iouClasses) * 100
+            csvSampleValues.append(f"{sampleAcc:.2f}")
+
+            with csvSamplesPath.open("a") as file:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:160]: `open` should only be called once. Opening the same file inside of for-loop and then only appending one value to it is very slow.

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    model.val(
+        project = valResPath,
+        data = yamlFilePath,
+        batch = taskRun.parameters["batchSize"],
+        imgsz = taskRun.parameters["imageSize"],
+        save_json = True,
+    )
+
+    jsonFiles = valResPath.rglob("predictions.json")
+    for path in jsonFiles:
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:201]: Why is for-loop used here? It seems redundant. `jsonFiles` returns a generator value which might, or might not have elements. The rest of the code will crash if no elements are present in `jsonFiles` generator. Properly handle cases where there are no elements, or more than 1 element yielded by `jsonFiles` generator.

File: tasks/image-segmentation-yolov8/src/validate.py [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+        jsonPath = path
+
+    with jsonPath.open("r") as file:
+        predictData = list(json.load(file))
+
+    csvSamplesPath = folder_manager.createTempFolder("csv_samples") / "sample_results.csv"
+
+    processPrediction(taskRun, predictData, csvSamplesPath)
+
+    accuracy = createCsvDatasetResults(taskRun, csvSamplesPath)
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/src/validate.py:211]: This should not return accuracy. `processPrediction` function should return full results of the validation. `createCsvDatasetResults` should take those results and then create `dataset_results.csv` file which contains the results on the dataset level and `sample_results.csv` file which contains the result on the sample level. Each of the samples should also have a separate file `{sample.id}.csv` which contains the results per label for that sample.

File: tasks/image-segmentation-yolov8/task.yaml [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+      will be performed; if false, only training will be performed
+    value: false
+    data_type: bool
+    required: true
+  - name: batchSize
+    description: Number of segments into which dataset is split during the training
+    value: 16
+    data_type: int
+    required: true
+  - name: excludedClasses
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/task.yaml:39]: `excludedClasses` should not be a shared parameter, it is a training-only parameter.

File: tasks/image-segmentation-yolov8/task.yaml [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+- name: outputs
+  params:
+  - name: model
+    description: Model trained by this Task
+    value: null
+    data_type: model
+    required: false
+- name: sharedParameters
+  params:
+  - name: validationMode
```
Comment (dule1322) [inline tasks/image-segmentation-yolov8/task.yaml:28]: `validation` is enough

**Review by dule1322 — CHANGES_REQUESTED** [review]
